### DISCUSSION TITLE
What is the best way to have dialect-specific models?

### DISCUSSION THREAD

**Discussion by bholland** [sqlalchemy/sqlalchemy#7326] [Usage Questions]
Hello all,

I am trying to deal with multiple dialects that support different attributes on the model. For example, snowflake does not support indexes so on the model, a line like:

     from sqlalchemy import Sequence as sa_sequence
     from sqlalchemy.ext.declarative import declarative_base
     BaseModel: Any = declarative_base()
     class MyModel(BaseModel):
         __tablename__ = "table"
         table_name = Column(String(MAX_INDEXED_STRING_LENGTH), index=True, unique=True, nullable=False)

becomes

     from sqlalchemy import Sequence as sa_sequence
     from sqlalchemy.ext.declarative import declarative_base
     BaseModel: Any = declarative_base()
     class MyModel(BaseModel):
         __tablename__ = "table"
         table_name  = Column(String(MAX_INDEXED_STRING_LENGTH), index=False, unique=True, nullable=False)

How can I accomplish this? I would love to have a solution where I don't have to have multiple model files.

**Comment by CaselIT** [comment]
Hi,

If you have an engine available at module definition time you can just define a function like:
```py
# file db.py
from or import environ
from sqlalchemy import create_engine
engine = create_engine(environ.get('DB_URI'))
def supports_index():
  return engine.dialect.name != 'snowflake'
```
then use it 
```py

 from sqlalchemy import Sequence as sa_sequence
 from sqlalchemy.ext.declarative import declarative_base
 from .db import  supports_index

 BaseModel: Any = declarative_base()
 class MyModel(BaseModel):
     __tablename__ = "table"
     table_name = Column(String(MAX_INDEXED_STRING_LENGTH), index=supports_index(), unique=True, nullable=False)
```

For types you can use [`with_variant`](https://docs.sqlalchemy.org/en/14/core/type_api.html?highlight=with_variant#sqlalchemy.types.TypeEngine.with_variant)

**Comment by zzzeek** [comment]
Hey there -

this is conditional DDL, so I would follow the approach at https://docs.sqlalchemy.org/en/14/core/ddl.html#controlling-ddl-sequences .  unfortunately for Index it's a little more verbose because we want to prevent the Index from being created at all even in Python, so it can work like this:

```py
from sqlalchemy import Column
from sqlalchemy import event
from sqlalchemy import Index
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.engine.create import create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.schema import CreateIndex
from sqlalchemy.schema import DropIndex

BaseModel = declarative_base()


class MyModel(BaseModel):
    __tablename__ = "table"
    id = Column(Integer, primary_key=True)
    table_name = Column(String(10), unique=True, nullable=False)


@event.listens_for(MyModel.__table__, "after_create")
def add_index_mymodel(*arg, **kw):
    return CreateIndex(
        Index("my_index", MyModel.__table__.c.table_name)
    ).execute_if(dialect="sqlite")(*arg, **kw)


e = create_engine("sqlite://", echo=True)
BaseModel.metadata.create_all(e)
```
